File: Projeto/src/br/uece/aps/classes/livro/Livro.java
package br.uece.aps.classes.livro;

import java.util.List;

public class Livro {
    private String titulo;
    private float preco;
    private float avaliacao;
    private float qntAvaliacoes;
    private Editora editora;
    private List<Autor> autores;
    private List<Categoria> categorias;
  
    
    public Livro(String titulo, float preco, Editora editora) {
    	this.titulo = titulo;
    	this.preco = preco;
    	this.editora = editora;
    	this.avaliacao = 0;
    	this.qntAvaliacoes = 0;
    }

	public String getTitulo() {
		return titulo;
	}

	public void setTitulo(String titulo) {
		this.titulo = titulo;
	}

	public float getPreco() {
		return preco;
	}

	public void setPreco(float preco) {
		this.preco = preco;
	}
	
	public void resetAvaliacao() {
		this.avaliacao = 0;
		this.qntAvaliacoes = 0;
	}
	
	public void updateAvaliacao(float avaliacao) {
		this.avaliacao *= this.qntAvaliacoes;
		this.avaliacao = (this.avaliacao + this.avaliacao) / ++this.qntAvaliacoes;
	}
	
	public String getEditora() {
		return editora.getEditora();
	}
	
	public void updateCategoria(Categoria categoria) {
		this.categorias.add(categoria);
	}
	
	public void updateCategoria(List<Categoria> categorias) {
		this.categorias = categorias;
	}
	
	public void removeCategoria(Categoria categoria) {
		int index = this.categorias.indexOf(categoria);
		if(index != -1) {
			categorias.remove(index);
		}
	}
	
	public void updateAutores(Autor autor) {
		this.autores.add(autor);
	}
	
	public void updateAutores(List<Autor> autores) {
		this.autores = autores;
	}
	
	public void removeAutor(Autor autor) {
		int index = this.autores.indexOf(autor);
		if(index != -1) {
			categorias.remove(index);
		}
	}
	
}

File: livraria_uece/lib/classes/livro/livro.dart

import 'package:livraria_uece/classes/livro/autor.dart';
import 'package:livraria_uece/classes/livro/categoria.dart';
import 'package:livraria_uece/classes/livro/editora.dart';

class Livro {
  String _id;
  List<String> _url_capa;
  String _titulo;
  double _preco;
  List<int> _avaliacao;
  Editora _editora;
  List<Autor> _autores;
  List<Categoria> _categorias;

  Livro({ String id, List<String> url_capa, String titulo, double preco, List<int> avaliacao, Editora editora, List<Autor> autores, List<Categoria> categorias }) {
    _id = id;
    _url_capa = url_capa;
    _titulo = titulo;
    _preco = preco;
    _avaliacao = avaliacao ?? new List();
    _editora = editora;
    _autores = autores ?? new List();
    _categorias = categorias ?? new List();
  }

  void newCategoria(Categoria categoria) {
    _categorias.add(categoria);
  }

  void removeCategoria(Categoria categoria) {
    _categorias.remove(categoria);
  }

  void newAutor(Autor autor) {
    _autores.add(autor);
  }

  void removeAutor(Autor autor) {
    _autores.remove(autor);
  }

  void newAvaliacao(var avaliacao) {
    _avaliacao.add(avaliacao);
  }

  double get avaliacao {
    double avaliacao = 0;
    _avaliacao.forEach((element) => avaliacao += element );
    return avaliacao / _avaliacao.length ;
  }

  String get id => _id;

  set id(String value) {
    _id = value;
  }

  List<String> get url_capa => _url_capa;

  set url_capa(List<String> value) {
    _url_capa = value;
  }

  String get titulo => _titulo;

  set titulo(String value) {
    _titulo = value;
  }

  double get preco => _preco;

  set preco(double value) {
    _preco = value;
  }

  Editora get editora => _editora;

  set editora(Editora value) {
    _editora = value;
  }

  List<Autor> get autores => _autores;

  set autores(List<Autor> value) {
    _autores = value;
  }

  List<Categoria> get categorias => _categorias;

  set categorias(List<Categoria> value) {
    _categorias = value;
  }

  String toString(){
    return "\n  # Livro "+_id.toString()+' '+_url_capa.toString()+' '+_titulo.toString()+' '+_preco.toString()+' '+_avaliacao.toString()
      +' editoras '+_editora.toString()+' autores '+_autores.toString()+' categorias '+_categorias.toString();
  }
}